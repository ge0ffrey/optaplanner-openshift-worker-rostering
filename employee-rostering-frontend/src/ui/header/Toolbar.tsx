/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Button,
  ButtonVariant,
  Dropdown,
  DropdownItem,
  DropdownToggle,
  Toolbar,
  ToolbarGroup,
  ToolbarItem,
} from '@patternfly/react-core';
import { BellIcon, CogIcon } from '@patternfly/react-icons';
import * as React from 'react';
import Tenant from '../../domain/Tenant';
import TenantRestServiceClient from '../../services/TenantRestServiceClient';
import { isUndefined } from "util";

interface ToolbarProps { }
interface ToolbarState {
  currentTenant?: Tenant;
  tenantList: Tenant[];
  isTenantSelectOpen: boolean;
}

class ToolbarComponent extends React.Component<ToolbarProps, ToolbarState> {
  tenantRestService: TenantRestServiceClient = new TenantRestServiceClient();
  toolbarState: ToolbarState;

  constructor(props: ToolbarProps) {
    super(props);
    this.toolbarState = { tenantList: [], isTenantSelectOpen: false };
    this.updateTenantList();
  }

  updateTenantList() {
    this.tenantRestService.getTenantList().then(tenantList => {
      if (tenantList.length > 0) {
        if (isUndefined(this.toolbarState.currentTenant) || !tenantList.includes(this.toolbarState.currentTenant)) {
          this.toolbarState = {
            currentTenant: tenantList[0],
            tenantList: tenantList,
            isTenantSelectOpen: this.toolbarState.isTenantSelectOpen
          };
        }
        else {
          this.toolbarState = {
            currentTenant: this.toolbarState.currentTenant,
            tenantList: tenantList,
            isTenantSelectOpen: this.toolbarState.isTenantSelectOpen
          };
        }
      }
      else {
        this.toolbarState = { tenantList: [], isTenantSelectOpen: this.toolbarState.isTenantSelectOpen };
      }
      this.refresh();
    });
  }

  setCurrentTenant(newTenantId: number) {
    this.toolbarState = {
      currentTenant: this.toolbarState.tenantList[newTenantId],
      tenantList: this.toolbarState.tenantList,
      isTenantSelectOpen: false
    };
    this.refresh();
  }

  setIsTenantSelectOpen(isOpen: boolean) {
    this.toolbarState = { currentTenant: this.toolbarState.currentTenant, tenantList: this.toolbarState.tenantList, isTenantSelectOpen: isOpen };
    this.refresh();
  }

  refresh() {
    this.setState(this.toolbarState);
  }

  render() {
    if (isUndefined(this.toolbarState.currentTenant)) {
      return <Toolbar><ToolbarGroup /> <ToolbarGroup>
        <ToolbarItem>
          <Button
            id="horizontal-example-uid-01"
            aria-label="Notifications actions"
            variant={ButtonVariant.plain}
          >
            <BellIcon />
          </Button>
        </ToolbarItem>
        <ToolbarItem>
          <Button
            id="horizontal-example-uid-02"
            aria-label="Settings actions"
            variant={ButtonVariant.plain}
          >
            <CogIcon />
          </Button>
        </ToolbarItem>
      </ToolbarGroup></Toolbar>
    }
    else {
      let { currentTenant, tenantList, isTenantSelectOpen } = this.toolbarState;
      return <Toolbar>
        <ToolbarGroup>
          <ToolbarItem>
            <Dropdown
              isPlain={true}
              position="right"
              // tslint:disable-next-line:no-console
              onSelect={event => this.setCurrentTenant(parseInt(event.currentTarget.id))}
              isOpen={isTenantSelectOpen}
              toggle={
                <DropdownToggle onToggle={() => this.setIsTenantSelectOpen(!isTenantSelectOpen)}>
                  {currentTenant.name}
                </DropdownToggle>}
              dropdownItems={tenantList.map((tenant, index) => {
                return <DropdownItem id={index.toString()} key={index}>{tenant.name}</DropdownItem>;
              })}
            />
          </ToolbarItem>
        </ToolbarGroup>
        <ToolbarGroup>
          <ToolbarItem>
            <Button
              id="horizontal-example-uid-01"
              aria-label="Notifications actions"
              variant={ButtonVariant.plain}
            >
              <BellIcon />
            </Button>
          </ToolbarItem>
          <ToolbarItem>
            <Button
              id="horizontal-example-uid-02"
              aria-label="Settings actions"
              variant={ButtonVariant.plain}
            >
              <CogIcon />
            </Button>
          </ToolbarItem>
        </ToolbarGroup>
      </Toolbar>
    }
  }
}

export default ToolbarComponent;
