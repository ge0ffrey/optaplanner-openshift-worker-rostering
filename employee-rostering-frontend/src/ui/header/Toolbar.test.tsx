/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { ToolbarComponent, Props } from './Toolbar';

describe('Toolbar Component', () => {
  it('should render correctly with no tenants', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...noTenants} />);
    expect(toJson(toolbarComponent)).toMatchSnapshot();
  });

  it('should render correctly with a few tenants', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...twoTenants} />);
    expect(toJson(toolbarComponent)).toMatchSnapshot();
  });

  it('should render correctly when tenant select is open', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...twoTenants} />);
    (toolbarComponent.instance() as ToolbarComponent).setIsTenantSelectOpen(true);
    expect(toJson(toolbarComponent)).toMatchSnapshot();
  });

  it('should refresh tenant list on mount', async () => {
    const toolbarComponent = shallow(<ToolbarComponent {...twoTenants} />);
    await toolbarComponent.instance().componentDidMount;
    expect(twoTenants.refreshTenantList).toBeCalled();
  });

  it('should call change tenant on tenant change', () => {
    const toolbarComponent = shallow(<ToolbarComponent {...twoTenants} />);
    (toolbarComponent.instance() as ToolbarComponent).setCurrentTenant(2);
    expect(twoTenants.changeTenant).toBeCalled();
    expect(twoTenants.changeTenant).toBeCalledWith(2);
  });
});

const noTenants: Props = {
  tenantList: [],
  currentTenantId: 0,
  refreshTenantList: jest.fn(),
  changeTenant: jest.fn()
};

const twoTenants: Props = {
  tenantList: [
    {
      id: 1,
      version: 0,
      name: "Tenant 1"
    },
    {
      id: 2,
      version: 0,
      name: "Tenant 2"
    }
  ],
  currentTenantId: 2,
  refreshTenantList: jest.fn(),
  changeTenant: jest.fn()
};