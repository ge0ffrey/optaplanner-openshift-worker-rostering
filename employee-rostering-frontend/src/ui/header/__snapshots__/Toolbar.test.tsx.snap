// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Component should render correctly when tenant select is open 1`] = `
<Toolbar
  className={null}
>
  <ToolbarGroup
    className={null}
  >
    <ToolbarItem
      className={null}
    >
      <Dropdown
        className=""
        direction="down"
        dropdownItems={
          Array [
            <Item
              className=""
              component="a"
              data-tenantid={1}
              href="#"
              isDisabled={false}
              isHovered={false}
            >
              Tenant 1
            </Item>,
            <Item
              className=""
              component="a"
              data-tenantid={2}
              href="#"
              isDisabled={false}
              isHovered={false}
            >
              Tenant 2
            </Item>,
          ]
        }
        isGrouped={false}
        isOpen={true}
        isPlain={true}
        onSelect={[Function]}
        position="right"
        toggle={
          <DropdownToggle
            className=""
            iconComponent={[Function]}
            id=""
            isActive={false}
            isFocused={false}
            isHovered={false}
            isOpen={false}
            isPlain={false}
            onToggle={[Function]}
            parentRef={null}
          >
            Tenant 2
          </DropdownToggle>
        }
      />
    </ToolbarItem>
  </ToolbarGroup>
  <ToolbarGroup
    className={null}
  >
    <ToolbarItem
      className={null}
    >
      <Button
        aria-label="Notifications actions"
        id="horizontal-example-uid-01"
        variant="plain"
      >
        <BellIcon
          color="currentColor"
          size="sm"
          title={null}
        />
      </Button>
    </ToolbarItem>
    <ToolbarItem
      className={null}
    >
      <Button
        aria-label="Settings actions"
        id="horizontal-example-uid-02"
        variant="plain"
      >
        <CogIcon
          color="currentColor"
          size="sm"
          title={null}
        />
      </Button>
    </ToolbarItem>
  </ToolbarGroup>
</Toolbar>
`;

exports[`Toolbar Component should render correctly with a few tenants 1`] = `
<Toolbar
  className={null}
>
  <ToolbarGroup
    className={null}
  >
    <ToolbarItem
      className={null}
    >
      <Dropdown
        className=""
        direction="down"
        dropdownItems={
          Array [
            <Item
              className=""
              component="a"
              data-tenantid={1}
              href="#"
              isDisabled={false}
              isHovered={false}
            >
              Tenant 1
            </Item>,
            <Item
              className=""
              component="a"
              data-tenantid={2}
              href="#"
              isDisabled={false}
              isHovered={false}
            >
              Tenant 2
            </Item>,
          ]
        }
        isGrouped={false}
        isOpen={false}
        isPlain={true}
        onSelect={[Function]}
        position="right"
        toggle={
          <DropdownToggle
            className=""
            iconComponent={[Function]}
            id=""
            isActive={false}
            isFocused={false}
            isHovered={false}
            isOpen={false}
            isPlain={false}
            onToggle={[Function]}
            parentRef={null}
          >
            Tenant 2
          </DropdownToggle>
        }
      />
    </ToolbarItem>
  </ToolbarGroup>
  <ToolbarGroup
    className={null}
  >
    <ToolbarItem
      className={null}
    >
      <Button
        aria-label="Notifications actions"
        id="horizontal-example-uid-01"
        variant="plain"
      >
        <BellIcon
          color="currentColor"
          size="sm"
          title={null}
        />
      </Button>
    </ToolbarItem>
    <ToolbarItem
      className={null}
    >
      <Button
        aria-label="Settings actions"
        id="horizontal-example-uid-02"
        variant="plain"
      >
        <CogIcon
          color="currentColor"
          size="sm"
          title={null}
        />
      </Button>
    </ToolbarItem>
  </ToolbarGroup>
</Toolbar>
`;

exports[`Toolbar Component should render correctly with no tenants 1`] = `
<Toolbar
  className={null}
>
  <ToolbarGroup
    className={null}
  />
  <ToolbarGroup
    className={null}
  >
    <ToolbarItem
      className={null}
    >
      <Button
        aria-label="Notifications actions"
        id="horizontal-example-uid-01"
        variant="plain"
      >
        <BellIcon
          color="currentColor"
          size="sm"
          title={null}
        />
      </Button>
    </ToolbarItem>
    <ToolbarItem
      className={null}
    >
      <Button
        aria-label="Settings actions"
        id="horizontal-example-uid-02"
        variant="plain"
      >
        <CogIcon
          color="currentColor"
          size="sm"
          title={null}
        />
      </Button>
    </ToolbarItem>
  </ToolbarGroup>
</Toolbar>
`;
